### DISCUSSION TITLE
Is there a need for two separate arrays for rules and messages in FormRequests?

### DISCUSSION THREAD

**Discussion by AbbasRam** [laravel/framework#51439] [Ideas]
Mainly I use FormRequests for validating requests because I think this is more clean. Also because All application I write needs to be in a language other then english so I should return error messages in a specific language. That's why I need FormRequest instead of `$request->validate()` function.

I don't know which one is more popular (But I assume `$request->validate()` is more popular because nearly all of the questions on the internet regarding form validation turns around `$request->validate()` function)

But in the FormRequest version you should have two things. First you must have a rules() function which returns an array. All keys are form input field and all values are the validation rules that must be present for given input.
Secondly, you should have a messages() method. 

> "This method should return an array of attribute / rule pairs and their corresponding error messages" [Laravel docs.](https://laravel.com/docs/11.x/validation#customizing-the-error-messages)

**Problem:**
But there is an overhead here. One should repeat every combination of attribute / rule pairs in their messages array. So if its the case, I think it would be more useful if we were able to use just one `messages()` method because the messages array contains all the information to deduce the validation rules and their corresponding form input fields.

**Suggestion:**
We may have a default version where we only use `rules()` method with default messages. Or we may use only `messages()` method with custom messages and validation rules in the same place.

**Comment by ghost** [comment]
Hi. When I use custom messages, I write there all rules in words in one value, even if only one fails.

**Reply by AbbasRam** [reply]
Can you give an example of it? I didn't quite understand what you mean.

**Reply by ghost** [reply]
```
class MyForm extends FormRequest
{
    /**
     * @inheritdoc
     */
    public function messages(): array
    {
        return ['field' => 'field is required and must be positive or 0 integer'];
        // instead of 
        return [
            'field.required' => 'field is required',
            'field.integer' => 'field must be integer',
            'field.min' => 'field must be positive or 0 integer',
         ];
    }

    /**
     * @inheritdoc
     */
    public function rules(): array
    {
        return ['field' => 'required|integer|min:0'];
    }
}
```

**Reply by AbbasRam** [reply]
Yeah I see. This is a step better than every combination of key=>values. But in case one wants to differentiate between error messages, this method has limited capabilities I think.

**Reply by ghost** [reply]
What is the point of rejecting the request multiple times in steps because the field is not valid?
If not valid, expose all rules for the user to send next time a valid field.

**Reply by AbbasRam** [reply]
For example in your given example you are using only 3 validation rules. But in my projects there are cases where I am using more than 10 validation rule for a field. So I am not able to use one generic error message for all errors at once.

**Comment by ghost** [comment]
@AbbasRam set the messages based on the required language on construct. Or create a form class for each language that you select via dependency injection with an interface.
